### PR TITLE
trace: Logger helper

### PR DESCRIPTION
**Pull request by keegancsmith**
Note: I am unsure on this API. Please be brutal/nitpicky in the feedback :)

I don't see many uses of log.WithTrace, but I think this is because it is clunky to use since Span or Context may be nil on a context. This introduces a helper so we can start logging against a span more easily. Here is an example of using it I want:

```go
logger := trace.WithLogger(ctx, s.Log).Scoped("...", "...").With(...
```

This seems like the best place to add the logger, since we need to use internal functions in the trace pkg (ie won't work in lib/log).

I updated the documentation to include an example of it. Right now the documentation has one other place it calls WithTrace, but it needs to be reworked to be more realistic. I think calling log.Scoped on something that is traced seems surprising to me.

Test Plan: unused, so just that CI is happy.

### CODE REVIEW

**Comment by keegancsmith** [comment]
This link might not survive due to PR updates, but here is an example of what inspired this PR https://github.com/sourcegraph/sourcegraph/pull/33783/files#diff-a80efa474c7214df1748a2d26bc3cea14fa73d37ea85df7e61421e2c9bc59971R29-R32

File: internal/trace/traceutil.go [hunk 1]
```diff
@@ -73,6 +74,15 @@ func ContextFromSpan(span opentracing.Span) *otfields.TraceContext {
 	return nil
 }
 
+// WithLogger will set the TraceContext on l if ctx has one. This is a
+// convenience function around l.WithLogger for the common case.
```
Comment (bobheadxi) [inline internal/trace/traceutil.go:78]:
```suggestion
// convenience function around l.WithTrace for the common case.
```

**Review by bobheadxi — APPROVED** [review]
This is awesome, much more ergonomic for sure 😁 Thank you!

File: doc/dev/how-to/add_logging.md [hunk 2]
```diff
@@ -107,8 +107,8 @@ func (w *Worker) DoSomething(params ...int) {
 If you are kicking off a long-running process, you can spawn a child logger and use it directly to maintain relevant context:
 
 ```go
-func (w *Worker) DoBigThing(params ...int) {
-  doLog := w.logger.WithTrace(/* ... */).With("params", params)
+func (w *Worker) DoBigThing(ctx context.Context, params ...int) {
+  doLog := trace.WithLogger(ctx, w.logger).With("params", params)
```
Comment (bobheadxi) [inline doc/dev/how-to/add_logging.md:111]: Another possible API:

```suggestion
  doLog := trace.Logger(ctx, w.logger).With("params", params)
```

semantically, "get the trace(d) logger"

**Review by burmudar — APPROVED** [review]

File: internal/trace/traceutil.go [hunk 3]
```diff
@@ -73,6 +74,15 @@ func ContextFromSpan(span opentracing.Span) *otfields.TraceContext {
 	return nil
 }
 
+// WithLogger will set the TraceContext on l if ctx has one. This is a
+// convenience function around l.WithTrace for the common case.
+func WithLogger(ctx context.Context, l sglog.Logger) sglog.Logger {
```
Comment (keegancsmith) [inline internal/trace/traceutil.go:79]:
```suggestion
// Logger will set the TraceContext on l if ctx has one. This is a
// convenience function around l.WithTrace for the common case.
func Logger(ctx context.Context, l sglog.Logger) sglog.Logger {
```

**Pull request merged**